cpmpy/expressions/variables.py

```diff
@@ -441,10 +441,23 @@ def __getitem__(self, index):
 
         # multi-dimensional index
         if isinstance(index, tuple) and any(isinstance(el, Expression) for el in index):
+
+            if len(index) != self.ndim:
+                raise NotImplementedError("CPMpy does not support returning an array from an Element constraint. Provide an index for each dimension. If you really need this, please report on github.")
+
             # find dimension of expression in index
-            expr_dim = next(dim for dim,idx in enumerate(index) if isinstance(idx, Expression))
-            arr = self[tuple(index[:expr_dim])] # select remaining dimensions
-            index = index[expr_dim:]
+            expr_dim = [dim for dim,idx in enumerate(index) if isinstance(idx, Expression)]
+            if len(expr_dim) == 1: # optimization, only 1 expression, reshape to 1d-element
+                # TODO can we do the same for more than one Expression? Not sure...
```

Comment:
should be possible to reduce the amount of dimensions by the amount of constants in index, but might be a bit of a braintwister to implement, already found this code hard to understand.

Reply:
Yes, we implemented it already but without altering the order of dimensions in the variables.
So `arr[1,x,y]` gets converted to `arr[1][x,y]` but `arr[x,y,1]` does not get changed